### PR TITLE
Tests/protocols test

### PR DESCRIPTION
**Pull request by Totoual**
## Proposed changes

Please comment on these merge but after discussing with Marco I'll revert the changes. The problem is that the error handling doesn't leave any room for extra tests. I modified some core scripts in order to be able to test.

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

DELETE INCLUSIVE THIS AND BELOW FOR STANDARD PR
------

## Release summary

Version number: [e.g. 1.0.1]

## Release details

Describe in short the main changes with the new release.

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [ ] I am making a pull request against the `master` branch (left side), from `develop`
- [ ] Lint and unit tests pass locally
- [ ] I built the documentation and updated it with the latest changes
- [ ] I've added an item in `HISTORY.rst` for this release
- [ ] I bumped the version number in the `aea/__version__.py` file.
- [ ] I bumped the version number in every Docker image of the repo and published it. Also, I built and published them with tag `latest` 
- [ ] I have checked that the documentation about the `aea cli` tool works

## Further comments

Write here any other comment about the release, if any.

### CODE REVIEW

File: tests/test_protocols/test_fipa.py [hunk 1]
```diff
@@ -66,3 +68,57 @@ def test_fipa_propose_serialization():
...
     assert p1[0].values == p2[0].values
     assert p1[1].values == p2[1].values
+
+
+def test_fipa_accept_serialization():
+    """Test that the serialization for the 'fipa' protocol works."""
+    msg = FIPAMessage(message_id=0, dialogue_id=0, target=0, performative=FIPAMessage.Performative.ACCEPT)
+    msg_bytes = FIPASerializer().encode(msg)
+    envelope = Envelope(to="receiver", sender="sender", protocol_id=FIPAMessage.protocol_id, message=msg_bytes)
+    envelope_bytes = envelope.encode()
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_fipa.py]: why are we testing envelope encoding here too?

File: tests/test_protocols/test_fipa.py [hunk 2]
```diff
@@ -66,3 +68,57 @@ def test_fipa_propose_serialization():
...
+    msg = FIPAMessage(
+        performative=FIPAMessage.Performative.ACCEPT,
+        message_id=0,
+        dialogue_id=0,
+        destination="publicKey",
+        target=1)
+
+    with mock.patch("aea.protocols.fipa.message.FIPAMessage.Performative") as mock_performative_enum:
+        mock_performative_enum.ACCEPT.value = "unknown"
+        assert not msg.check_consistency()
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_fipa.py]: please add a message for assert failure

File: tests/test_protocols/test_fipa.py [hunk 3]
```diff
@@ -66,3 +68,57 @@ def test_fipa_propose_serialization():
...
+    """Test the string value of the performatives."""
+    assert str(FIPAMessage.Performative.CFP) == "cfp"
+    assert str(FIPAMessage.Performative.PROPOSE) == "propose"
+    assert str(FIPAMessage.Performative.DECLINE) == "decline"
+    assert str(FIPAMessage.Performative.ACCEPT) == "accept"
+    assert str(FIPAMessage.Performative.MATCH_ACCEPT) == "match_accept"
+
+
+def test_serialisation_fipa():
+    """Tests the serialisation of a fipa message."""
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_fipa.py]: it is not clear from the test description what you expect to happen

Comment (DavidMinarsch) [reply tests/test_protocols/test_fipa.py]: please remove comments

File: tests/test_connections/test_oef/test_oef_serializer.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+from aea.protocols.oef.serialization import OEFSerializer
+from aea.protocols.oef.models import Attribute, DataModel, Description
+
+
+def test_oef_serialization():
+    """Testing the serialization of the OEF."""
+    foo_datamodel = DataModel("foo", [Attribute("bar", int, True, "A bar attribute.")])
+    desc = Description({"bar": 1}, data_model=foo_datamodel)
+    msg = OEFMessage(oef_type=OEFMessage.Type.REGISTER_SERVICE, id=1, service_description=desc, service_id="")
+    msg_bytes = OEFSerializer().encode(msg)
```
Comment (DavidMinarsch) [inline tests/test_connections/test_oef/test_oef_serializer.py]: I think encode, then decode would be a better test.

File: tests/test_connections/test_oef/test_models.py [hunk 5]
```diff
@@ -54,6 +54,7 @@ def test_description(self):
         attribute_bar = Attribute("bar", str, True, "a bar attribute.")
         data_model_foobar = DataModel("foobar", [attribute_foo, attribute_bar], "A foobar data model.")
         description_foobar = Description({"foo": 1, "bar": "baz"}, data_model=data_model_foobar)
+        description_foobar.__iter__()
```
Comment (DavidMinarsch) [inline tests/test_connections/test_oef/test_models.py]: why is this needed?

File: aea/protocols/scaffold/message.py [hunk 6]
```diff
@@ -38,21 +38,21 @@ def __str__(self):
...
-        assert self.check_consistency(), "MyScaffoldMessage initialization inconsistent."
+        self.consistent = self.check_consistency()
+#       assert self.check_consistency(), "MyScaffoldMessage initialization inconsistent."
 
     def check_consistency(self) -> bool:
         """Check that the data is consistent."""
         try:
             raise NotImplementedError
-        except (AssertionError, ValueError):
+        except (AssertionError, ValueError, NotImplementedError):
```
Comment (DavidMinarsch) [inline aea/protocols/scaffold/message.py]: I think we don't want the consistency check to pass with False on NotImplementedError. I think we want it to throw the error.

File: aea/protocols/scaffold/message.py [hunk 7]
```diff
@@ -38,21 +38,21 @@ def __str__(self):
...
-                 **kwargs):
+    def __init__(self, oef_type: Optional[Type] = None, **kwargs):
         """
         Initialize.
 
         :param oef_type: the type of message.
         """
         super().__init__(type=oef_type, **kwargs)
-        assert self.check_consistency(), "MyScaffoldMessage initialization inconsistent."
+        self.consistent = self.check_consistency()
```
Comment (DavidMinarsch) [inline aea/protocols/scaffold/message.py]: what is the rational behind this change?

File: aea/protocols/fipa/message.py [hunk 8]
```diff
@@ -62,6 +63,7 @@ def __init__(self, message_id: Optional[int] = None,
                          target=target,
                          performative=FIPAMessage.Performative(performative),
                          **kwargs)
+#        self.consistent = self.check_consistency()
```
Comment (DavidMinarsch) [inline aea/protocols/fipa/message.py]: please remove this comment

File: aea/protocols/fipa/message.py [hunk 9]
```diff
@@ -39,6 +39,7 @@ class Performative(Enum):
         ACCEPT = "accept"
         MATCH_ACCEPT = "match_accept"
         DECLINE = "decline"
+        UNKNOWN = "unknown"
```
Comment (DavidMinarsch) [inline aea/protocols/fipa/message.py]: please remove this performative

File: tests/test_protocols/test_fipa.py (hunk 1 above)
Comment (Totoual) [reply tests/test_protocols/test_fipa.py]: Because I keep my tests in different functions, I don't want to mess with @marcofavorito 's code.

File: tests/test_connections/test_oef/test_models.py (hunk 5 above)
Comment (Totoual) [reply tests/test_connections/test_oef/test_models.py]: I manually call this in order to check that returns a description. I probably should test it in some different way that I don't know yet. I tried to change it to `myDescription = iter(description_foobar)` but I get an error.

File: aea/protocols/oef/serialization.py [hunk 10]
```diff
@@ -90,7 +90,8 @@ def decode(self, obj: bytes) -> Message:
             new_body["agents"] = list(json_msg["agents"])
         elif oef_type in {OEFMessage.Type.OEF_ERROR}:
             operation = json_msg["operation"]
```
Comment (DavidMinarsch) [inline aea/protocols/oef/serialization.py]: Ok I see that this is necessary because of line 64 where the operation is cast to a string. Is casting to a string in line 64 necessary? Otherwise we could just not cast to strong and your fix would not be needed.

File: tests/test_connections/test_oef/test_models.py [hunk 11]
```diff
@@ -54,6 +54,7 @@ def test_description(self):
         attribute_bar = Attribute("bar", str, True, "a bar attribute.")
         data_model_foobar = DataModel("foobar", [attribute_foo, attribute_bar], "A foobar data model.")
         description_foobar = Description({"foo": 1, "bar": "baz"}, data_model=data_model_foobar)
+#        my_description = iter(description_foobar)
         oef_description = OEFObjectTranslator.to_oef_description(description_foobar)
```
Comment (DavidMinarsch) [inline tests/test_connections/test_oef/test_models.py]: If this line of commented code. Is not needed then delete please

File: tests/test_protocols/test_default.py [hunk 12]
```diff
@@ -38,3 +51,24 @@ def test_default_error_serialization():
...
+        with mock.patch("aea.protocols.default.message.DefaultMessage.Type")\
+                as mock_type_enum:
+            mock_type_enum.BYTES.value = "unknown"
+            body = {}  # Dict[str, Any]
+            msg_type = DefaultMessage.Type(msg.get("type"))
+            body["type"] = str(msg_type.value)
+            content = cast(bytes, msg.get("content"))
+            body["content"] = base64.b64encode(content).decode("utf-8")
+            bytes_msg = json.dumps(body).encode("utf-8")
+            assert DefaultSerializer().decode(bytes_msg), "Type is recognized!"
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_default.py]: I think it would be better to assert equality here

File: tests/test_protocols/test_fipa.py (hunk 1 above)
Comment (DavidMinarsch) [reply tests/test_protocols/test_fipa.py]: If you are saying that envelope encoding is only tested here then please move it to the correct test file. Otherwise remove the envelope encoding test. It’s ok to add to @marcofavorito code.

File: tests/test_protocols/test_oef_message.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        "The string representation is different than oef_error"
+    assert str(OEFMessage.Type.DIALOGUE_ERROR) == "dialogue_error",\
+        "The string representation is different than dialogue_error"
+    assert str(OEFMessage.Type.SEARCH_RESULT) == "search_result",\
+        "The string representation is different than search_result"
+
+
+def test_oef_message_consistency():
+    """Tests the consistency of an OEFMessage."""
+    foo_datamodel = DataModel("foo", [Attribute("bar", int,
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_oef_message.py]: Your tests are not named particularly well. It’s hard to understand from the test of the name whether it is testing for failure or success. Please be more explicit.

Comment (Totoual) [reply tests/test_protocols/test_oef_message.py]: But it tests both. I have an assert that returns TRUE (this message is correct) ` assert msg.check_consistency(), "We expect the consistency to return TRUE" ` and I also have an assert that returns FALSE `assert not msg.check_consistency(), "Expect the consistency to return False" `. I am testing the message consistency in general.

**Comment by DavidMinarsch** [comment]
Fixes #145
